dashboard/app/models/user.rb

```diff
@@ -742,13 +742,16 @@ def self.name_from_omniauth(raw_name)
 
   CLEVER_ADMIN_USER_TYPES = ['district_admin', 'school_admin'].freeze
   def self.from_omniauth(auth, params, session = nil)
-    omniauth_user = find_by_credential(type: auth.provider, id: auth.uid)
-
-    unless omniauth_user
-      omniauth_user = create do |user|
-        initialize_new_oauth_user(user, auth, params)
+    # Make sure no matching user account can be created by another thread
+    # between trying to find the user and creating a new one
+    transaction do
```

Comment:
this doesn't look right to me. you're treating this like a mutex, which would say, "atomically look up the user or create one if the user was not found". but that's not what transaction does, rather if anything inside the transaction fails then all mutations made by the transaction will be undone. so, this won't prevent 2 threads from both entering the transaction block, looking up the user, and then trying to create the user. and because the create doesn't fail, both transactions will succeed.

I believe we usually deal with this by having a unique constraint at the DB layer, which we then rely upon to make sure the second create request fails.

Reply:
Oh, my understanding was that transactions _did_ act as a mutex by locking the table; and that in fact that was why we should avoid casually sprinkling them around.

We do have the option of making the second create request fail, but what I'd really like to do is to ensure that the second create request will always be able to find the user created by the first, which as I understand will require some kind of lock. Do we have a specific desire to avoid mutexes in our codebase?